src/pbrt/util/taggedptr.h

```diff
@@ -743,9 +743,14 @@ class TaggedPointer {
     template <typename T>
     PBRT_CPU_GPU TaggedPointer(T *ptr) {
         uint64_t iptr = reinterpret_cast<uint64_t>(ptr);
-        DCHECK_EQ(iptr & ptrMask, iptr);
         constexpr unsigned int type = TypeIndex<T>();
+#if defined(PBRT_IS_WINDOWS) && defined(__HIPCC__)
+        uint64_t aptr = (iptr & tagMask) >> 5ull;
+        bits = (iptr & ptrMask) | aptr | ((uint64_t)type << (tagShift + 2ull));
```

Comment:
It's a little unclear to me what's going on with the tricky pointer tagging stuff here. You're generally using `tagShift + 2` here for HIPCC, though have decreased `tagShift` by 2 when it's defined below. Can this be cleaned up? (I'm not sure if there are other places where you're depending on it having a smaller value though.)

Reply:
Actually, the HIP pointers uses internally the top bits for some (to me unknown purposes); this shifts PBRT tags to lower bits to avoid conflicts.

Reply:
Makes sense. Generally speaking, let's just try to centralize the bit operations in a small number of well-commented blocks of code, since it's all pretty tricky stuff.